### PR TITLE
[Sparse] Add SpMM and SDDMM.

### LINKED ISSUE #4855: [Sparse] Support sparse-dense matmul and SDDMM
## 🔨Work Item

Project tracker: https://github.com/orgs/dmlc/projects/2

### PR DESCRIPTION
**Pull request by czkkkkkk**
## Description
Resolve #4855.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/mock_sparse2/matmul.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    A1: SparseMatrix, A2: Union[torch.Tensor, SparseMatrix, DiagMatrix]
+) -> Union[torch.Tensor, SparseMatrix]:
+    """Internal function for multiplying a sparse matrix by a dense/sparse/diagonal matrix
+
+    Parameters
+    ----------
+    A1 : SparseMatrix
+        Matrix of shape (N, M), with values of shape (nnz1)
+    A2 : torch.Tensor, SparseMatrix, or DiagMatrix
+        Matrix of shape (M, P). If it is a SparseMatrix or DiagMatrix,
```
Comment (mufeili) [inline python/dgl/mock_sparse2/matmul.py:70]: Do you support 1D `torch.Tensor`?

File: dgl_sparse/src/sparse_format.cc [hunk 2]
```diff
@@ -92,5 +92,20 @@ std::shared_ptr<COO> COOTranspose(const std::shared_ptr<COO>& coo) {
   return COOFromOldDGLCOO(dgl_coo_tr);
 }
 
+HeteroGraphPtr COOToDGLGraph(const std::shared_ptr<COO>& coo) {
+  auto dgl_coo = COOToOldDGLCOO(coo);
```
Comment (mufeili) [inline dgl_sparse/src/sparse_format.cc:96]: Leave a comment for the meaning of `2`?

File: dgl_sparse/include/sparse/sddmm.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+// clang-format on
+
+#include <sparse/sparse_matrix.h>
+#include <torch/script.h>
+
+namespace dgl {
+namespace sparse {
+
+/**
+ * @brief Performs a sampled matrix multiplication of a sparse matrix and two
```
Comment (mufeili) [inline dgl_sparse/include/sparse/sddmm.h:20]:
```suggestion
 * @brief Perform a sampled matrix multiplication of a sparse matrix and two
```

File: dgl_sparse/include/sparse/sddmm.h [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ * dense matrix. For efficiency, `mat2_tr` is the transposition of the matrix to
```
Comment (mufeili) [inline dgl_sparse/include/sparse/sddmm.h:21]:
```suggestion
 * dense matrices. For efficiency, `mat2_tr` is the transposition of the matrix to
```

File: dgl_sparse/include/sparse/sddmm.h [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+ * be multiplied. If the sparse matrix has shape (n, m), `mat1` and `mat2_tr`
+ * must have shapes of `(n, k)` and `(m, k)` or
+ * `(n,)` and `(m,)` respectively. And the returned tensor has shape
+ * `(sparse_matrix->nnz())`.
+ *
+ * This function does not take care of autograd.
+ *
+ * @param sparse_matrix The sparse matrix.
```
Comment (mufeili) [inline dgl_sparse/include/sparse/sddmm.h:29]:
```suggestion
 * @param sparse_mat The sparse matrix.
```

File: dgl_sparse/include/sparse/spmm.h [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+// clang-format on
+
+#include <sparse/sparse_matrix.h>
+#include <torch/script.h>
+
+namespace dgl {
+namespace sparse {
+
+/**
+ * @brief Performs a matrix multiplication of the sparse matrix and dense
```
Comment (mufeili) [inline dgl_sparse/include/sparse/spmm.h:20]:
```suggestion
 * @brief Perform a matrix multiplication of the sparse matrix and dense
```

File: dgl_sparse/include/sparse/spmm.h [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+ * matrix. It uses the sparse formats of `sparse_mat` and non-zero values of
+ * `sparse_val` for SpMM. The `sparse_val` must be 1-dimensional. If the sparse
+ * matrix has shape (n, m), the dense matrix must have shape (m, k) or (m,). and
+ * the returned dense matrix has shape (n, k) or (n,
+ * ).
+ *
+ * This function does not take care of autograd.
+ *
+ * @param sparse_matrix The sparse matrix.
```
Comment (mufeili) [inline dgl_sparse/include/sparse/spmm.h:29]:
```suggestion
 * @param sparse_mat The sparse matrix.
```

File: dgl_sparse/include/sparse/spmm.h [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+ * @param dense_mat The dense matrix.
+ *
+ * @return Dense tensor.
+ */
+torch::Tensor SpMMImpl(
+    const c10::intrusive_ptr<SparseMatrix>& sparse_mat,
+    torch::Tensor sparse_val, torch::Tensor dense_mat);
+
+/**
+ * @brief Performs a matrix multiplication of the sparse matrix and dense
```
Comment (mufeili) [inline dgl_sparse/include/sparse/spmm.h:40]:
```suggestion
 * @brief Perform a matrix multiplication of the sparse matrix and dense
```

File: dgl_sparse/include/sparse/spmm.h [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+ * matrix. The sparse matrix must have 1-dimensional values. If the sparse
+ * matrix has shape (n, m), the dense matrix must have shape (m, k) or (m,). and
```
Comment (mufeili) [inline dgl_sparse/include/sparse/spmm.h:42]:
```suggestion
 * matrix has shape (n, m), the dense matrix must have shape (m, k) or (m,), and
```

File: dgl_sparse/include/sparse/spmm.h [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+ * the returned dense matrix has shape (n, k) or (n,).
+ *
+ * This function supports autograd for both the sparse and dense matrix.
+ *
+ * @param sparse_mat The sparse matrix.
+ * @param dense_mat The dense matrix.
+ *
+ * @return Result dense matrix.
```
Comment (mufeili) [inline dgl_sparse/include/sparse/spmm.h:50]:
```suggestion
 * @return Dense matrix.
```

File: include/dgl/kernel.h [hunk 11]
```diff
@@ -42,10 +46,16 @@ void SpMM(
  * @param ufeat The source node feature.
  * @param vfeat The destination node feature.
  * @param out The output feature on edge.
+ * @param lhs_target Type of `ufeat`. (0: source; 1: edge; 2: destination)
+ * @param rhs_target Type of `vfeat`. (0: source; 1: edge; 2: destination)
  */
 void SDDMM(
-    const std::string& op, HeteroGraphPtr graph, NDArray ufeat, NDArray efeat,
```
Comment (mufeili) [inline include/dgl/kernel.h:47]: Does this mean that the use of `efeat` was a mistake?

File: python/dgl/mock_sparse2/matmul.py (hunk 1 above)
Comment (czkkkkkk) [reply python/dgl/mock_sparse2/matmul.py:70]: Yes.

File: include/dgl/kernel.h (hunk 11 above)
Comment (czkkkkkk) [reply include/dgl/kernel.h:47]: Yes. It should be `vfeat` as in its definition code.

File: python/dgl/mock_sparse2/matmul.py (hunk 1 above)
Comment (mufeili) [reply python/dgl/mock_sparse2/matmul.py:70]: Then, you need to update the doc.

**Review by mufeili — APPROVED** [review]
I'm good other than some minor comments.

**Pull request merged**